usaspending_api/database_scripts/matview_generator/matview_sql_generator.py

```diff
@@ -0,0 +1,211 @@
+        "unique": true,
+        "method": "<method>"
+    }
+}
+'''
+
+TEMPLATE = {
+    'create_matview': 'CREATE MATERIALIZED VIEW {} AS\n{};',
+    'drop_matview': 'DROP MATERIALIZED VIEW IF EXISTS {};',
+    'rename_matview': 'ALTER MATERIALIZED VIEW {}{} RENAME TO {};',
+    'cluster_matview': 'CLUSTER VERBOSE {} USING {};',
+    'vacuum': 'VACUUM ANALYZE VERBOSE {};',
+    'create_index': 'CREATE {}INDEX {} ON {} USING {}({}){}{};',
+    'rename_index': 'ALTER INDEX {}{} RENAME TO {};',
+}
+HEADER = [
+    '--------------------------------------------------------',
+    '-- Created using matview_sql_generator.py             --',
+    '--    The SQL definition is stored in a json file     --',
+    '--    Look in matview_generator for the code.         --',
+    '--                                                    --',
+    '--  DO NOT DIRECTLY EDIT THIS FILE!!!                 --',
+    '--------------------------------------------------------',
+]
+MAX_NAME_LENGTH = 45  # postgres max 63 ascii chars
+RANDOM_CHARS = str(uuid4())[:8]
+CLUSTERING_INDEX = None
+DEST_FOLDER = '../matviews/'
+OVERWRITE_FILE = True
+
+
+def ingest_json(path):
+    with open(path) as f:
+        doc = json.load(f)
+    return doc
+
+
+def create_index_string(matview_name, index_name, idx):
+    if idx.get('cluster_on_this', False):
+        global CLUSTERING_INDEX
+        CLUSTERING_INDEX = index_name
+    idx_method = idx.get('method', 'BTREE')  # if missing, defaults to BTREE
+    idx_unique = 'UNIQUE ' if idx.get('unique', False) else ''
+    idx_where = ' WHERE ' + idx['where'] if idx.get('where', None) else ''
+    idx_with = ''
+    if idx_method.upper() == 'BTREE':
+        idx_with = ' WITH (fillfactor = 100)'  # reduce btree index size by 10%
+
+    idx_cols = []
+    for col in idx['columns']:
+        index_def = [col["name"]]  # Critial to have col or expression. Exception if missing
+        if col.get('order', None):  # if missing, skip and let postgres default to ASC
+            index_def.append(col['order'])
+        if col.get('collation', None):  # if missing, skip and let postgres use default
+            index_def.append('COLLATE ' + col['collation'])
+        if col.get('opclass', None):  # if missing, skip and let postgres choose
+            index_def.append(col['opclass'])
+        idx_cols.append(' '.join(index_def))
+    idx_str = TEMPLATE['create_index'].format(
+        idx_unique,
+        index_name,
+        matview_name,
+        idx_method,
+        ', '.join(idx_cols),
+        idx_with,
+        idx_where,
+    )
+    return idx_str
+
+
+def create_sql_strings(sql_json):
+    ''' Desired ordering of steps for final SQL:
+        1. Drop existing "_temp" and "_old" matviews
+        2. Create new matview
+        3. Create indexes for new matview
+        4. (optional) Cluster matview on index
+        5. vacuum analyze verbose <matview>
+        6. Rename existing matview, append with _old
```

Comment:
At the top of every script the _old matviews are dropped

Reply:
one steady-state goal would be to manually drop the old matviews once rudimentary checks on the new matviews are successful. the drop materialized views on the top of the files is in case the temp or old matviews are still around